diff -up gnome-session-2.23.6/compat/at-spi-registryd-wrapper.c.crash gnome-session-2.23.6/compat/at-spi-registryd-wrapper.c
--- gnome-session-2.23.6/compat/at-spi-registryd-wrapper.c.crash	2008-08-12 10:07:48.000000000 -0400
+++ gnome-session-2.23.6/compat/at-spi-registryd-wrapper.c	2008-08-12 10:10:15.000000000 -0400
@@ -98,7 +98,7 @@ set_gtk_modules (DBusGProxy *gsm)
                 modules[i] = NULL;
 
                 value = g_strjoinv (":", modules);
-                g_strfreev (modules);
+                g_free (modules);
         } else {
                 value = g_strdup ("gail:atk-bridge");
         }
