diff -up gnome-session-2.23.92/gnome-session/gsm-manager.c.empty-string-leak gnome-session-2.23.92/gnome-session/gsm-manager.c
--- gnome-session-2.23.92/gnome-session/gsm-manager.c.empty-string-leak	2008-09-14 21:44:18.000000000 -0400
+++ gnome-session-2.23.92/gnome-session/gsm-manager.c	2008-09-14 21:48:06.000000000 -0400
@@ -913,6 +913,7 @@ _on_query_end_session_timeout (GsmManage
                 app_id = g_strdup (gsm_client_peek_app_id (l->data));
                 if (IS_STRING_EMPTY (app_id)) {
                         /* XSMP clients don't give us an app id unless we start them */
+                        g_free (app_id);
                         app_id = gsm_client_get_app_name (l->data);
                 }
 
@@ -1529,6 +1530,7 @@ on_client_end_session_response (GsmClien
                 app_id = g_strdup (gsm_client_peek_app_id (client));
                 if (IS_STRING_EMPTY (app_id)) {
                         /* XSMP clients don't give us an app id unless we start them */
+                        g_free (app_id);
                         app_id = gsm_client_get_app_name (client);
                 }
 
