diff -up gnome-session-3.3.2/data/hardware-compatibility.jx gnome-session-3.3.2/data/hardware-compatibility
--- gnome-session-3.3.2/data/hardware-compatibility.jx	2011-12-13 14:50:39.000000000 -0500
+++ gnome-session-3.3.2/data/hardware-compatibility	2011-12-13 14:54:24.325533866 -0500
@@ -15,6 +15,9 @@
 # Intel 830-865
 -Intel\(R\) 8[[:digit:]]{2,2}[^[:digit:]]
 
+# Pre-R300 radeon
+-Mesa DRI R[12]00
+
 # Old Mesa software GL renderer
 -software rasterizer
 
