--- gnome-session-2.5.90/splash/splash-widget.c.icons	2004-02-24 14:13:22.000000000 +0000
+++ gnome-session-2.5.90/splash/splash-widget.c	2004-02-24 14:14:24.000000000 +0000
@@ -46,7 +46,11 @@
 	{ N_("The Panel"),               "gnome-panel",           "gnome-panel" },
 	{ N_("Session Manager Proxy"),   "gnome-smproxy",         "gnome-session" },
 	{ N_("Nautilus"),                "nautilus",              "gnome-fs-desktop" },
-	{ N_("Desktop Settings"),        "gnome-settings-daemon", "gnome-settings" }
+	{ N_("Desktop Settings"),        "gnome-settings-daemon", "gnome-settings" },
+	{ N_("Authentication Icon"),     "pam-panel-icon",        "keyring.png" },
+	{ N_("CD Device Handler"),       "magicdev",              "gnome-dev-cdrom.png" },
+	{ N_("Red Hat Network Monitor"), "rhn-applet-gui",        "up2date.png" },
+	{ N_("Print Monitor"),           "eggcups",               "printer.png" }
 };
 
 static const SplashApp *
