--- gnome-session-2.23.5/gnome-session/gsm-manager.c.orig	2008-07-21 15:50:16.000000000 -0400
+++ gnome-session-2.23.5/gnome-session/gsm-manager.c	2008-07-21 15:51:18.000000000 -0400
@@ -1612,7 +1612,11 @@
                         continue;
                 }
 
-                desktop_file = g_strdup_printf ("%s.desktop", (char *) a->data);
+                if (strcmp (a->data, "nautilus") == 0) {
+                        desktop_file = g_strdup ("gnome-nautilus.desktop");
+                } else {
+                        desktop_file = g_strdup_printf ("%s.desktop", (char *) a->data);
+                }
 
                 g_debug ("GsmManager: Looking for: %s", desktop_file);
 
