diff -up gnome-session-2.25.92/gnome-session/main.c.debug-spew gnome-session-2.25.92/gnome-session/main.c
--- gnome-session-2.25.92/gnome-session/main.c.debug-spew	2009-03-06 12:10:21.724791954 -0500
+++ gnome-session-2.25.92/gnome-session/main.c	2009-03-06 12:10:46.640936065 -0500
@@ -57,7 +57,7 @@
 
 static gboolean failsafe = FALSE;
 static gboolean show_version = FALSE;
-static gboolean debug = TRUE;
+static gboolean debug = FALSE;
 
 static void
 on_bus_name_lost (DBusGProxy *bus_proxy,
