--- gnome-session-2.17.5/data/gnome-session.schemas.in.window-manager	2007-01-08 17:32:28.000000000 -0500
+++ gnome-session-2.17.5/data/gnome-session.schemas.in	2007-01-11 14:05:30.000000000 -0500
@@ -2,11 +2,22 @@
 <gconfschemafile>
    <schemalist>
       <schema>
+         <key>/schemas/apps/gnome-session/rh/window_manager</key>
+         <applyto>/apps/gnome-session/rh/window_manager</applyto>
+         <owner>gnome</owner>
+         <type>string</type>
+         <default>metacity</default>
+         <locale name="C">
+	    <short>Window Manager to start</short>
+	    <long>The Window manager that gnome-session will start. Valid values are "metacity" and "compiz".</long>
+         </locale>
+      </schema>
+      <schema>
          <key>/schemas/apps/gnome-session/options/show_splash_screen</key>
          <applyto>/apps/gnome-session/options/show_splash_screen</applyto>
          <owner>gnome</owner>
          <type>bool</type>
-         <default>true</default>
+         <default>false</default>
          <locale name="C">
             <short>Show the splash screen</short>
             <long>Show the splash screen when the session starts up</long>
--- gnome-session-2.17.5/gnome-session/gnome-wm.window-manager	2007-01-08 17:32:28.000000000 -0500
+++ gnome-session-2.17.5/gnome-session/gnome-wm	2007-01-11 14:08:11.000000000 -0500
@@ -3,7 +3,9 @@
 # The user can specify his prefered WM by setting the WINDOW_MANAGER
 # environment variable.
 #
-# If this is not set, we search a list of known windowmanagers and use
+# If this is not set, we check the /apps/gnome-session/rh/window_manager key
+# and go off that.  finally, if all else fails we search a list of known 
+# windowmanagers and use
 # the first one that is found in the users's PATH
 #
 
@@ -38,6 +40,9 @@
 done
 
 # WINDOW_MANAGER overrides all
+if [ -z "$WINDOW_MANAGER" ] ; then
+	WINDOW_MANAGER=$(gconftool-2 --get /apps/gnome-session/rh/window_manager)
+fi
 
 if [ -z "$WINDOW_MANAGER" ] ; then
   # Create a list of window manager we can handle, trying to only use the
