Index: gnome-session/gsm-sound.c
===================================================================
--- gnome-session/gsm-sound.c	(revision 4457)
+++ gnome-session/gsm-sound.c	(working copy)
@@ -12,27 +12,12 @@
 #include <libgnome/gnome-sound.h>
 #include <libgnome/gnome-triggers.h>
 
-#define ENABLE_SOUND_KEY        "/desktop/gnome/sound/enable_esd"
 #define ENABLE_EVENT_SOUNDS_KEY "/desktop/gnome/sound/event_sounds"
 
 static gboolean
 sound_enabled (void)
 {
-  GConfClient *client;
-  GError      *error = NULL;
-  gboolean     retval;
-
-  client = gsm_get_conf_client ();
-
-  retval = gconf_client_get_bool (client, ENABLE_SOUND_KEY, &error);
-  if (error)
-    {
-      g_warning ("Error getting value of " ENABLE_SOUND_KEY ": %s", error->message);
-      g_error_free (error);
-      return FALSE;  /* Fallback value */
-    }
-
-  return retval;
+  return TRUE;
 }
 
 static gboolean
