diff -up gnome-session-3.3.5/gnome-session/gsm-system.c.interface gnome-session-3.3.5/gnome-session/gsm-system.c
--- gnome-session-3.3.5/gnome-session/gsm-system.c.interface	2012-02-09 23:35:03.634312123 -0500
+++ gnome-session-3.3.5/gnome-session/gsm-system.c	2012-02-09 23:35:12.438311807 -0500
@@ -34,7 +34,7 @@ enum {
 
 static guint signals[LAST_SIGNAL] = { 0 };
 
-G_DEFINE_INTERFACE (GsmSystem, gsm_system, G_TYPE_INTERFACE);
+G_DEFINE_INTERFACE (GsmSystem, gsm_system, G_TYPE_OBJECT)
 
 static void
 gsm_system_default_init (GsmSystemInterface *iface)
