diff -up gnome-session-2.19.92/gnome-session/ice.c.ice-leaks gnome-session-2.19.92/gnome-session/ice.c
--- gnome-session-2.19.92/gnome-session/ice.c.ice-leaks	2007-09-11 01:56:57.000000000 -0400
+++ gnome-session-2.19.92/gnome-session/ice.c	2007-09-11 01:57:49.000000000 -0400
@@ -455,10 +455,13 @@ clean_ice (void)
   /* During shutdown we don't want to report errors via a dialog.  */
   write_authfile (authfile, entries, FALSE);
 
+  g_slist_foreach (entries, (GFunc)IceFreeAuthFileEntry, NULL);
   g_slist_free (entries);
 
   g_free (input_id);
   IceFreeListenObjs (num_sockets, sockets);
+
+  g_slist_foreach (auth_entries, (GFunc)IceFreeAuthFileEntry, NULL);
   g_slist_free (auth_entries);
   
   ice_depth = ~0;	/* We are very frozen, like totally off */
