diff -up gnome-session-2.23.92/gnome-session/gsm-inhibit-dialog.c.inhibitor-leak gnome-session-2.23.92/gnome-session/gsm-inhibit-dialog.c
--- gnome-session-2.23.92/gnome-session/gsm-inhibit-dialog.c.inhibitor-leak	2008-09-14 19:48:46.000000000 -0400
+++ gnome-session-2.23.92/gnome-session/gsm-inhibit-dialog.c	2008-09-14 19:47:48.000000000 -0400
@@ -496,6 +496,7 @@ add_inhibitor (GsmInhibitDialog *dialog,
 
         /* FIXME: get info from xid */
 
+        desktop_file = NULL;
         name = NULL;
         pixbuf = NULL;
         freeme = NULL;
@@ -602,6 +603,9 @@ add_inhibitor (GsmInhibitDialog *dialog,
         if (pixbuf != NULL) {
                 g_object_unref (pixbuf);
         }
+        if (desktop_file != NULL) {
+                egg_desktop_file_free (desktop_file);
+        }
 }
 
 static gboolean
