diff -up gnome-session-2.23.92/gnome-session/gsm-xsmp-server.c.ListenObj-leak gnome-session-2.23.92/gnome-session/gsm-xsmp-server.c
--- gnome-session-2.23.92/gnome-session/gsm-xsmp-server.c.ListenObj-leak	2008-09-14 13:36:09.000000000 -0400
+++ gnome-session-2.23.92/gnome-session/gsm-xsmp-server.c	2008-09-14 13:38:00.000000000 -0400
@@ -606,6 +606,9 @@ gsm_xsmp_server_finalize (GObject *objec
 
         g_return_if_fail (xsmp_server->priv != NULL);
 
+        IceFreeListenObjs (xsmp_server->priv->num_xsmp_sockets, 
+                           xsmp_server->priv->xsmp_sockets);
+
         if (xsmp_server->priv->client_store != NULL) {
                 g_object_unref (xsmp_server->priv->client_store);
         }
